### PR TITLE
Proto exercises

### PR DESCRIPTION
**Pull request by elParcero**

### CODE REVIEW

File: exercises/03_key_value_header.py [hunk 1]
```diff
@@ -36,18 +36,28 @@
...
+i0_data = np.array(df.i0, dtype="float")
+
 #plotting
+plt.ion()
 plt.figure(0)
 plt.clf()
-
-np.log(df.it/df.i0)
-
+plt.plot(np.log(it_data/i0_data), color='#be0119')  #scarlet color
```
Comment (jrmlhermitte) [inline exercises/03_key_value_header.py]: great! for a final step, can you plot versus energy?
xlabel should then be "energy (keV)"

File: exercises/04-reading-raw-data.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 '''
 import numpy as np 
 import os
+import pandas as pd
```
Comment (jrmlhermitte) [inline exercises/04-reading-raw-data.py]: add this near beginning:
```py3
'''
Some helper functions. We don't have to worry about the details.
These are hardware level details.
Use these to apply to the 'counts' column for an files etc
'''
fc = 7.62939453125e-05
adc2counts = lambda x: ((int(x, 16) >> 8) - 0x40000) * fc \
        if (int(x, 16) >> 8) > 0x1FFFF else (int(x, 16) >> 8)*fc
enc2counts = lambda x: int(x) if int(x) <= 0 else -(int(x) ^ 0xffffff - 1)


```

File: exercises/04-reading-raw-data.py [hunk 3]
```diff
@@ -20,43 +21,70 @@
...
+#print the sum of list
+for i in range(len(sum_of_time)):
+	print('Sum of total time in {0} = {1}s'.format(an_files_only[i], sum_of_time[i]))
+
+#will read in each file, store contents in pandas dataframe, and store in dict where data[filename] = dataframe
+data = dict()
+index = 0
+
+for file in an_files_only:
+	df = pd.read_csv((directory_string + "/" + an_files_only[index]) , delimiter=" ", names = ['time (s)', 'time (ns)', 'index', 'counts'], header = None)
```
Comment (jrmlhermitte) [inline exercises/04-reading-raw-data.py]: add these lines:
```py3
    df['volts'] = df['counts'].apply(adc2counts)
    df['total time (s)'] = df['time (s)'] + 1e-9*df['time (ns)']

```

File: exercises/04-reading-raw-data.py [hunk 4]
```diff
@@ -20,43 +21,70 @@
...
+	data[file] = df
+	index += 1
+
+print("\n")
+df['total time (s)'] = df['time (s)'] + 1e-9*df['time (ns)']
```
Comment (jrmlhermitte) [inline exercises/04-reading-raw-data.py]: remove this line since it would be added above

File: exercises/03_key_value_header.py (hunk 1 above)
Comment (elParcero) [reply exercises/03_key_value_header.py]: would it be like this?
'''python
energy_data = np.array(df.energy, dtype="float")

plt.figure(1)
plt.clf()
plt.plot( ((np.log(it_data/i0_data)) , energy_data))
plt.show()
'''

**Comment by elParcero** [comment]
You want the readings for the an_files, and en_files to be coded into a function?

File: exercises/03_key_value_header.py (hunk 1 above)
Comment (jrmlhermitte) [reply exercises/03_key_value_header.py]: (use the back tick symbol not the quote for code blocks)

almost, `energy_data` should be first

File: exercises/04-reading-raw-data.py (hunk 4 above)
Comment (jrmlhermitte) [reply exercises/04-reading-raw-data.py]: I saw that (`data[file] = df`). looks good!

**Comment by jrmlhermitte** [comment]
yeah, something like:
```py3
def read_an_file(filename):
    # returns a data frame

an_data[file] = read_an_file_filename)
```

etc

File: exercises/03_key_value_header.py (hunk 1 above)
Comment (elParcero) [reply exercises/03_key_value_header.py]: I'll give it a try, with the back tick symbol
```python
plt.plot(energy_data, np.log(it_data/i0_data), color='#be0119')  #scarlet color
```

Comment (elParcero) [reply exercises/03_key_value_header.py]: It worked, thanks !

**Comment by elParcero** [comment]
Alright, I made a commit in branch : exercise-4 , let me know if that is okay.

**Pull request merged**